### PR TITLE
Fix import hints breaking node paths in animations

### LINKED ISSUE #38163: Exported animations in glTF 2 from Blender don't works in Godot (for parented object and object with argument like -col)
**Godot version: v3.2.2.beta1.officla**

**OS/device including version: windows 8.1**

Issue description:
Exported animations in glTF 2 from Blender in some conditions don't works in Godot.

To be more precise, the animations stop to works when there is parenting between objects or -col or -colonly argument in the name.

Seem to be a mismatch between the name in the Track of the AnimationPlayer and the name of the node.
When parenting object in Blender, the name in the Track is "object_name/parented_object_name" but the node name in Godot is "object_name"

Same problem with argument like -col.
The name in the Track of AnimationPlayer keep the -col but the node when importing is renamed without the -col argument.

Some screenshots to illustrate:
screenshot-tree
screenshot-animationplayer

### LINKED ISSUE #38809: Using import hints for glTF 2.0 files breaks animation references
**Godot version:**
3.2.1.stable.official

**OS/device including version:**
Windows 10 Home v1903

**Issue description:**
When importing a glTF 2.0 file, Godot will recognise import hints to objects such as '-col', adding features such as a collision node and removing the hint, so for example 'Marker-col' turns 'Marker'. However, the animation references aren't changed, so you get an error such as:

> Node not found: Marker-col.
> On Animation: 'Marker-loop', couldn't resolve track:  'Marker-col'.

**Steps to reproduce:**

Create a .glb file with an object with a hint, and animations involving that object. Import it into Godot, open it, click on the AnimationPlayer, and the error will occur and the animations will not work.

**Minimal reproduction project:**

[Import Hint Bug Report Example.zip](https://github.com/godotengine/godot/files/4640972/Import.Hint.Bug.Report.Example.zip)

### LINKED ISSUE #42041: Importer doesn't drop -col and -colonly suffixes in animation tracks, but removes them from node names
When importing any .dae or .gltf file exported using Blender's (2.90) built-in exporters, the animations don't work for -col/-colonly objects,

The reason appears to be quite simple: the importer drops the -col and -colonly suffix from the node names, but the animation tracks are referring to (non-existing) nodes with -col and -colonly suffix.

(Side note: when exporting gltf with Blender 2.90, "Group by NLA track" needs to be checked off, otherwise, only one of the existing tracks appear in theAnimationPlayer node, possibly a separate issue for gltf importer)

### PR DESCRIPTION
**Pull request by rburing**
Keep track of renamed nodes (which are processed first) so that in the animations (which are processed last) the paths can be fixed.

Fixes #38809. Importing the minimal example there now gives the following verbose output:

    Fix: Renamed Marker-col to Marker
    Fix: Correcting node path in animation track: Marker-col should be Marker

Small bonus: fixes the unreported bug that the `convcolonly` hint was not removed in case of an empty object with a draw type.

Bugsquad Edit:
- Fixes https://github.com/godotengine/godot/issues/42041 
- Fixes https://github.com/godotengine/godot/issues/38163

### CODE REVIEW

**Comment by fire** [comment]
Slow to get to this, putting it on my todo list.

**Comment by fire** [comment]
https://github.com/godotengine/godot-tests/tree/master/tests/blend_export/blend

Can you make a test case pr with the new tests?

**Comment by rburing** [comment]
To clarify the (small) scope of this PR, let me note that the code that deals with empties is barely touched here (only fixing one renaming oversight). There remains a separate issue with them, namely importing them as custom collision shapes (based on the draw type) only works with files exported from Blender using ol' Better Collada: https://github.com/godotengine/godot-docs/issues/4374

**Comment by fire** [comment]
I added 43044-marker.blend to https://github.com/fire/godot-tests/tree/43044/tests/blend_export/blend .

Currently building the engine to test that it breaks in 4.0 standard and works in your pr.

**Comment by fire** [comment]
I couldn't see a difference, maybe I'm misunderstanding how this works. Notice I have a mesh called marker next to marker-col.

This PR. 
image

Master.

image

**Comment by rburing** [comment]
The original `marker.blend` file in #38809 contains only `Marker-col`; the `43044-marker.blend` test file contains a second `Marker` in addition to `Marker-col`. This reveals the following (hidden?) functionality: https://github.com/godotengine/godot/blob/c56a071d0c6e282dbb82aa7e69691e33f6c2fd7e/editor/import/resource_importer_scene.cpp#L514-L515
Namely, the `-col` suffix is not removed if it clashes with an existing name. In that case there is no need to fix animations, so the animations from the `43044-marker.blend` work even on `master` without modifications (and in my PR, the new code is simply not triggered, and so it also works). When there is no clash with an existing name (as will be the common case, and as it was in #38809), there will be a rename that removes the `-col` suffix, and the code in my PR will be triggered, fixing the animations.

So, `43044-marker.blend` currently does not test the code in this PR (the original `marker.blend` does).

**Comment by fire** [comment]
I wonder why Marker-col and not Marker2? That seems odd.

**Comment by rburing** [comment]
The quoted `if` statement completely prevents the rename (in case that there would be a clash), so in that sense it is normal.

If `Marker2` is preferred then the logic should be changed.

**Comment by reduz** [comment]
I will be reworking this code in the coming weeks, can probably use it as a reference but suggest review for 3.2

**Comment by fire** [comment]
What is the current state of this? Need some help to retest master.

**Comment by reduz** [comment]
This looks good, but likely needs a rebase. Any volunteers?

**Comment by rburing** [comment]
Thanks for the feedback. I'll rebase today.

File: editor/import/resource_importer_scene.cpp [hunk 1]
```diff
@@ -293,16 +293,17 @@ static void _gen_shape_list(const Ref<Mesh> &mesh, List<Ref<Shape3D>> &r_shape_l
 	}
 }
 
-Node *ResourceImporterScene::_fix_node(Node *p_node, Node *p_root, Map<Ref<Mesh>, List<Ref<Shape3D>>> &collision_map, LightBakeMode p_light_bake_mode) {
+Node *ResourceImporterScene::_fix_node(Node *p_node, Node *p_root, Map<Ref<Mesh>, List<Ref<Shape3D>>> &collision_map, LightBakeMode p_light_bake_mode, List<Pair<NodePath, Node *>> &p_node_renames) {
```
Comment (akien-mga) [inline editor/import/resource_importer_scene.cpp:296]: This should actually be named `r_node_renames` as it's actually mutated by the function (we use `r_` prefix to indicated that it's a mutable reference or "return" argument).

**Comment by rburing** [comment]
Rebased it is.

**Comment by rburing** [comment]
Originally this PR was doing the renaming work in the same loop as the one where tracks are being removed if they reference nodes with a `noimp` (no import) hint. That original version would have introduced a bug if an animation would have a track referencing a node with a `noimp` hint *and* a track referencing a node with an import hint that would trigger a rename. In the latest version the removing and renaming are done in two separate loops (first removing, then renaming), for simplicity and safety.

**Comment by rburing** [comment]
I added a local variable `bool nodes_were_renamed = r_node_renames.size() != 0`, so that we try to fix node paths (which involves iterating over all tracks of all animations) only when `nodes_were_renamed`.

If the original PR looked good and the (small) improvements are agreeable, then does it mean a positive review? :)

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
This might warrant a dedicated backport PR for `3.x` if needed there, as the change is not trivial to cherry-pick.

**Pull request merged**